src/data/api-service.ts

```diff
@@ -0,0 +1,154 @@
+import axios, { AxiosRequestConfig, AxiosResponse } from 'axios'
+
+import { warn } from '~/utils/console'
+import { AUDIO, IMAGE } from '~/constants/media'
+
+const DEFAULT_REQUEST_TIMEOUT = 30000
+
+/**
+ * Returns a slug with trailing slash for a given resource name.
+ * For media types, converts the name into resource slug when necessary (i.e. pluralizes 'image'),
+ * for other resources uses the resource name as the slug.
+ * @param resource - the first part of the request path
+ */
+export const getResourceSlug = (resource: string): string => {
+  const slug = { [AUDIO]: 'audio', [IMAGE]: 'images' }[resource] ?? resource
+  return `${slug}/`
+}
+/**
+ * @param errorCondition - if true, the `message` warning is logged in the console
+ * @param message - message to display if there is an error in request
+ * @param config - Axios config object that containing request url
+ */
+const validateRequest = (
+  errorCondition: boolean,
+  message: string,
+  config: AxiosRequestConfig
+): void => {
+  if (errorCondition) {
+    warn(
+      `There is a problem with the request url: ${message}.
+Please check the url: ${config.baseURL}${config.url}`
+    )
+  }
+}
+
+/**
+ *
+ * @param baseUrl - the optional base url to override the `env.apiUrl`
+ */
+export const createApiService = (baseUrl = process.env.apiUrl) => {
+  const client = axios.create({
+    baseURL: baseUrl,
+    timeout: DEFAULT_REQUEST_TIMEOUT,
+  })
+  client.interceptors.request.use(function (config) {
+    validateRequest(
+      !config.url?.endsWith('/'),
+      'API request urls should have a trailing slash',
+      config
+    )
+    validateRequest(
+      config.url?.includes('//') ?? false,
+      'API request urls should not have two slashes',
+      config
+    )
+    return config
+  })
+  client.interceptors.response.use(
+    (response) => response,
+    (error) => {
+      if (error.code === 'ECONNABORTED') {
+        return Promise.reject({
+          message: `timeout of ${
+            DEFAULT_REQUEST_TIMEOUT / 1000
+          } seconds exceeded`,
+          ...error,
+        })
+      }
+      return Promise.reject(error)
+    }
+  )
+
+  return {
+    /**
+     * @param resource - The endpoint of the resource
+     * @param params - Url parameter object
+     * @returns response  The API response object
+     */
+    query<T>(
```

Comment:
`T` defaulted to `unknown` before to make it easier to call from JavaScript (because in JS you can't pass a generic type parameter). It'd be nice to keep it that way if possible to allow for casting the inner response time later in the code rather than immediately when calling the methods.